src/backend/distributed/worker/worker_partition_protocol.c

```diff
@@ -719,87 +701,100 @@ CitusRemoveDirectory(const char *filename)
+		 * Iterate over all its contents and for each content, recurse into
+		 * this function. Also, make sure that we do not recurse into symbolic
+		 * links.
 		 */
-		if (S_ISDIR(fileStat.st_mode) && !FileIsLink(filename, fileStat))
-		{
-			const char *directoryName = filename;
-
-			DIR *directory = AllocateDir(directoryName);
-			if (directory == NULL)
-			{
-				ereport(ERROR, (errcode_for_file_access(),
-								errmsg("could not open directory \"%s\": %m",
-									   directoryName)));
-			}
-
-			StringInfo fullFilename = makeStringInfo();
-			struct dirent *directoryEntry = ReadDir(directory, directoryName);
-			for (; directoryEntry != NULL; directoryEntry = ReadDir(directory,
-																	directoryName))
-			{
-				const char *baseFilename = directoryEntry->d_name;
+		const char *directoryName = filename;
 
-				/* if system file, skip it */
-				if (strncmp(baseFilename, ".", MAXPGPATH) == 0 ||
-					strncmp(baseFilename, "..", MAXPGPATH) == 0)
-				{
-					continue;
-				}
-
-				resetStringInfo(fullFilename);
-				appendStringInfo(fullFilename, "%s/%s", directoryName, baseFilename);
-
-				CitusRemoveDirectory(fullFilename->data);
-			}
-
-			FreeStringInfo(fullFilename);
-			FreeDir(directory);
+		DIR *directory = AllocateDir(directoryName);
+		if (errno == ENOENT)
+		{
+			/* If directory was removed from under us we're done */
+			return;
+		}
+		if (errno == ENOTDIR)
+		{
+			/* If directory changed to a file underneath us, loop again to remove it with unlink */
+			continue;
 		}
 
-		/* we now have an empty directory or a regular file, remove it */
-		if (S_ISDIR(fileStat.st_mode))
+		/* ReadDir handles other errno from AllocateDir */
+		struct dirent *directoryEntry = ReadDir(directory, directoryName);
+		StringInfo fullFilename = makeStringInfo();
+		for (; directoryEntry != NULL; directoryEntry = ReadDir(directory,
+																directoryName))
 		{
-			removed = rmdir(filename);
+			const char *baseFilename = directoryEntry->d_name;
 
-			if (errno == ENOTEMPTY || errno == EEXIST)
+			/* if system file, skip it */
+			if (strncmp(baseFilename, ".", MAXPGPATH) == 0 ||
+				strncmp(baseFilename, "..", MAXPGPATH) == 0)
 			{
 				continue;
 			}
-		}
-		else
-		{
-			removed = unlink(filename);
-		}
 
-		if (removed != 0 && errno != ENOENT)
-		{
-			ereport(ERROR, (errcode_for_file_access(),
-							errmsg("could not remove file \"%s\": %m", filename)));
+			resetStringInfo(fullFilename);
+			appendStringInfo(fullFilename, "%s/%s", directoryName, baseFilename);
+
+			CitusRemoveDirectory(fullFilename->data);
 		}
 
-		return;
+		FreeStringInfo(fullFilename);
+		FreeDir(directory);
+	}
+}
+
+
+/*
+ * TryUnlink returns true if file is removed and false if the file is a
+ * directory. If an error occurs trying to remove the file this function calls
+ * ereport.
+ */
+static bool
+TryUnlink(const char *filename)
+{
+	int exitcode = unlink(filename);
+	if (exitcode == 0 || errno == ENOENT)
+	{
+		return true;
```

Comment:
Would it make sense to check EISDIR first? Future callers might be confused about the behaviour of this function.